### PR TITLE
Self-Recharge Modules

### PR DESCRIPTION
**Pull request by RustingWithYou**
Adds self-recharge modules. These can be attached to non-modular energy weapons which are not already self-charging, and will allow them to do what the name suggests. There are three types, with varying recharge speed - with slow and standard modules being available to research. The latter takes higher tech levels, as well as resources from mining. Fast modules are attached to specific guns associated with high-tech factions (specifically the Tupkala guns, Vaurca guns and pulse weapons). 

Several existing energy guns used by military factions have been given standard self-charge modules. These are:
- Blaster weapons, as they are used by the TCFL/potentially the TCAF until we add better guns for those guys.
- Hegemony laser weapons
- Non-Tup Skrell weapons

The expedition shotguns have also been given slow charge modules, to make their use more viable on expeditions.

The recharge rate of the advanced energy gun has been increased, to make it still an improvement in the field of self-charging guns over just sticking a module on a laser rifle. 

A standard self-charging module is purchasable from the antagonist uplink.

### CODE REVIEW

**Comment by RustingWithYou** [comment]
!review

File: code/datums/uplink/devices_and_tools.dm [hunk 1]
```diff
@@ -250,3 +250,10 @@
 	telecrystal_cost = 2
 	bluecrystal_cost = 2
 	path = /obj/item/stack/liquidbags/half_full
+
+/datum/uplink_item/tools/self_charge_module
```
Comment (alsoandanswer) [inline code/datums/uplink/devices_and_tools.dm:254]:
```suggestion
/datum/uplink_item/item/tools/self_charge_module
```

File: code/datums/uplink/devices_and_tools.dm [hunk 2]
```diff
@@ -250,3 +250,10 @@
 	telecrystal_cost = 2
 	bluecrystal_cost = 2
 	path = /obj/item/stack/liquidbags/half_full
+
+/datum/uplink_item/item/tools/self_charge_module
+	name = "Energy Weapon Recharging Module"
+	desc = "Attach this to an energy weapon to make it self-charging"
```
Comment (NonQueueingMatt) [inline code/datums/uplink/devices_and_tools.dm:256]:
```suggestion
	desc = "Attach this to an energy weapon to make it self-charging."
```

File: code/modules/projectiles/guns/energy.dm [hunk 3]
```diff
@@ -30,6 +30,8 @@
 	var/use_external_power = 0 //if set, the weapon will look for an external power source to draw from, otherwise it recharges magically
 	var/recharge_time = 4
 	var/charge_tick = 0
+	var/obj/item/device/self_charge_module/charge_module = null
```
Comment (NonQueueingMatt) [inline code/modules/projectiles/guns/energy.dm:33]:
```suggestion
	var/obj/item/device/self_charge_module/charge_module
```

File: code/modules/projectiles/self_charge.dm [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+/obj/item/device/self_charge_module/afterattack(atom/target, mob/user, proximity_flag)
+	if(proximity_flag)
+		if(istype(target, /obj/item/gun/energy))
+			var/obj/item/gun/energy/G = target
+			if(!G.accepts_charge_module)
+				to_chat(user, SPAN_NOTICE("\The [G] has nowhere to attach a recharging module!"))
+				return
+			if(!G.charge_module)
+				gun_insert(user, G)
+				to_chat(user, SPAN_NOTICE("You attach [src] to \the [G]"))
```
Comment (NonQueueingMatt) [inline code/modules/projectiles/self_charge.dm:28]:
```suggestion
				to_chat(user, SPAN_NOTICE("You attach \the [src] to \the [G]."))
```

File: code/modules/projectiles/self_charge.dm [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+				to_chat(user, SPAN_NOTICE("\The [G] already has a recharging module installed."))
+
+/obj/item/device/self_charge_module/proc/gun_insert(mob/living/user, obj/item/gun/energy/G)
+	gun = G
+	user.drop_from_inventory(src,gun)
+	gun.charge_module = src
+	gun.self_recharge = TRUE
+	gun.use_external_power = FALSE
+	gun.recharge_time = src.charge_rate
+	return
```
Comment (NonQueueingMatt) [inline code/modules/projectiles/self_charge.dm:39]:
```suggestion

```
no need to return at the end of a proc

**Review by NonQueueingMatt — COMMENTED** [review]
We really need a generic attachment system for guns at this point.

**Review by NonQueueingMatt — APPROVED** [review]

File: code/modules/projectiles/guns/energy.dm [hunk 6]
```diff
@@ -76,12 +78,28 @@
...
+	if(charge_module)
+		self_recharge = TRUE
+		use_external_power = FALSE
+		recharge_time = charge_module.charge_rate
 	update_maptext()
 
 /obj/item/gun/energy/Destroy()
 	QDEL_NULL(power_supply)
+	if(istype(charge_module))
+		qdel(charge_module)
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:90]:
```suggestion
		QDEL_NULL(charge_module)
```

File: code/modules/projectiles/self_charge.dm [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	name = "weapon recharging module"
+	desc = "A small device attached to the power cell of an energy weapon, designed to allow it to continue operation when away from a convenient recharger. This one seems outdated, with an abysmally slow recharge rate."
+	icon = 'icons/obj/charge_modules.dmi'
+	icon_state = "low-selfchargemodule"
+	origin_tech = list(TECH_POWER = 3, TECH_COMBAT = 3)
+	flags = CONDUCT
+	w_class = ITEMSIZE_TINY
+	drop_sound = 'sound/items/drop/component.ogg'
+	pickup_sound = 'sound/items/pickup/component.ogg'
+	var/obj/item/gun/energy/gun
```
Comment (FluffyGhoster) [inline code/modules/projectiles/self_charge.dm:11]: You need to define a Destroy proc where you null this reference if the module is destroyed

File: code/modules/projectiles/guns/energy.dm [hunk 8]
```diff
@@ -76,12 +78,28 @@
...
 	return ..()
 
+/obj/item/gun/energy/attackby(obj/item/I, mob/user)
+	. = ..()
+	if(charge_module && I.iswirecutter())
+		to_chat(user, SPAN_NOTICE("You snip the wires connecting [charge_module] to \the [src]'s battery'."))
+		charge_module.forceMove(get_turf(src))
+		user.put_in_hands(charge_module)
+		charge_module = null
+		self_recharge = FALSE
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:100]: You need to call a proc to the module itself to drop the reference to the gun once disconnected, see below

File: code/modules/projectiles/self_charge.dm [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+				to_chat(user, SPAN_NOTICE("\The [G] already has a recharging module installed."))
+
+/obj/item/device/self_charge_module/proc/gun_insert(mob/living/user, obj/item/gun/energy/G)
+	gun = G
+	user.drop_from_inventory(src,gun)
+	gun.charge_module = src
+	gun.self_recharge = TRUE
+	gun.use_external_power = FALSE
+	gun.recharge_time = src.charge_rate
+
```
Comment (FluffyGhoster) [inline code/modules/projectiles/self_charge.dm:40]: You need a gun_remove() proc here, where you null the reference to the gun and whatever else needs to happen, specific to the recharge module itself, if the module is removed from the gun

The gun should call this, see above

**Review by FluffyGhoster — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/self_charge.dm (hunk 9 above)
Comment (RustingWithYou) [reply code/modules/projectiles/self_charge.dm:40]: addressed

File: code/modules/projectiles/guns/energy.dm (hunk 8 above)
Comment (RustingWithYou) [reply code/modules/projectiles/guns/energy.dm]: addressed

File: code/modules/projectiles/self_charge.dm (hunk 7 above)
Comment (RustingWithYou) [reply code/modules/projectiles/self_charge.dm:11]: addressed

File: code/modules/projectiles/self_charge.dm [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	gun = null
+	src.forceMove(get_turf(user))
+	user.put_in_hands(src)
+	gun.self_recharge = FALSE
+	return
+
+/obj/item/device/self_charge_module/Destroy()
+	if(gun)
+		gun.charge_module = null
+	return . = ..()
```
Comment (FluffyGhoster) [inline code/modules/projectiles/self_charge.dm:52]:
```suggestion
	. = ..()
```
`.` is implicitly returned at the end of any proc, no need to use the return keyword

File: code/modules/projectiles/guns/energy.dm [hunk 11]
```diff
@@ -76,12 +78,25 @@
...
+	if(charge_module)
+		self_recharge = TRUE
+		use_external_power = FALSE
+		recharge_time = charge_module.charge_rate
 	update_maptext()
 
 /obj/item/gun/energy/Destroy()
 	QDEL_NULL(power_supply)
+	if(istype(charge_module))
+		QDEL_NULL(charge_module)
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:90]:
```suggestion
	QDEL_NULL(charge_module)
```
QDEL_NULL takes care to check if the charge_module is installed or not, no need to typecheck it before, and of course it shouldn't ref anything but a charge module in the variable

**Comment by NonQueueingMatt** [comment]
Unwanted.

**Pull request closed without merging**